### ISSUE TITLE
No ability to adjust ZSTD_parameters when using a ZSTD_CDict

### ISSUE THREAD

**Issue by indygreg** [facebook/zstd#358]
Let me preface this issue by saying I'm not sure if this is a valid request since it may not be compatible with the internals of zstd.

Creating a `ZSTD_CDict` either through `ZSTD_createCDict()` or `ZSTD_createCDict_advanced()` results in a `ZSTD_CCtx` being created inside the opaque `ZSTD_CDict` instance. The `ZSTD_CCtx` is derived from `ZSTD_parameters`. Put another way, a `ZSTD_CDict` is forever associated with a constant `ZSTD_parameters` instance.

`ZSTD_parameters` can be influenced by the size of the thing being compressed (e.g. `ZSTD_adjustCParams()` and `ZSTD_getCParams()`). When using `ZSTD_CDict`, the `ZSTD_parameters` are essentially associated with only a single source size. If you are reusing a `ZSTD_CDict` and the size of the inputs varies, you are unable to adjust `ZSTD_parameters` to account for this, possibly leading to using sub-optimal `ZSTD_parameters`.

It feels like there needs to be an API that allows subsequent users of a `ZSTD_CDict` to adjust `ZSTD_parameters` accordingly from varying input sizes.

Currently, you /might/ be able to do this by redefining `ZSTD_CCtx_s` so its internal `ZSTD_CCtx` is exposed. Then you can `ZSTD_copyCCtx()` to copy the `ZSTD_CCtx` with the initialized dictionary data and then use `ZSTD_compress_advanced()` with empty dictionary arguments but a custom `ZSTD_parameters` derived from the new input size and it will still use the dict data from the copied `ZSTD_CCtx`. That's hacky of course. And I'm not even sure it works.

Of course, the appropriate answer to this issue might be something along the lines of "the compression parameters adjustment based on size isn't relevant when using dictionary compression." I'm just trying to figure out when it is appropriate to cache and reuse a `ZSTD_CDict`.

**Comment by Cyan4973** [comment]
This is a pretty good in-depth investigation,
and you are completely right about the limitation.

Neither `ZSTD_createCDict()` nor `ZSTD_createCDict_advanced()` allow providing future source sizes.  As you mentioned, such source sizes can vary, so no choice can be "universal". Instead, these functions "bet" that future data to compress will be "rather small".

This makes sense in the context where `CDict` is created :
the only reason for this interface to exist is to get faster processing on small files,
by making the initialization process much faster.

Large files don't benefit much from dictionary, plus their longer processing time make the initialization delay a smaller issue for them. So they are not targeted by `CDict`.

This method has a price to pay : compression parameters are static for all future files compressed with the same `CDict`. If parameters were to be changed, the initialization process could no longer be shared. It would become necessary to init a new `CDict`.

Bottom line :
for the compression process to adapt dynamically to each message, use `ZSTD_compress_usingDict()`, which will re-initialize the context at each message with optimal parameter.

Using `CDict` is meant to remove the initialization stage. It has to accept that parameters will be the same for all future compression using the same `CDict`.

More complex, you could create multiple `CDict` with different compression parameters, and switch between them depending on application-controlled heuristics.

**Comment by Cyan4973** [comment]
Does this answer your question ?

**Comment by indygreg** [comment]
Yes, this answers my question.
